### PR TITLE
Refactors cli/node to have an independent urfave implementation

### PR DESCRIPTION
**Pull request by nkcr**
This change cleans up a bit the cli/node package. 
I extracted the urfave implementation to ease the creation of other CLIs, like the one I'm preparing to use the crypto package.

### CODE REVIEW

File: cli/urfave/mod.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
+// Package urfave provides a cli builder implementation based on the urfave/cli
+// library.
+package urfave
```
Comment (pierluca) [inline cli/urfave/mod.go:3]: why name this after urfave, considering that all the exported types, functions, etc. are NOT urfave-dependent? As a matter of fact, if I understand correctly, from this package's consumer's perspective, this package's dependence on urfave is an implementation detail.
We could swap it out for another dependency and the user wouldn't notice.

could we name the package after its function? would `builder` be suited ?

File: cli/urfave/mod.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		Flags:    buildFlags(b.flags),
+	}
+
+	app.Setup()
+
+	return app
+}
+
+// SetCommand implements cli.Builder.
+func (b *Builder) SetCommand(name string) cli.CommandBuilder {
```
Comment (pierluca) [inline cli/urfave/mod.go:49]: I'm a tiny bit conflicted.
In Go, I believe the best practice is to take abstractions as parameters and return concretions, i.e. take in interfaces and return structs. On the other hand, it makes a lot of sense to return that abstraction here, in terms of code eloquence...

File: cli/urfave/mod.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+func (b *cmdBuilder) SetAction(action cli.Action) {
+	b.action = action
+}
+
+// SetSubCommand implements cli.CommandBuilder.
+func (b *cmdBuilder) SetSubCommand(name string) cli.CommandBuilder {
+	builder := &cmdBuilder{
+		subcommands: make(map[string]*cmdBuilder),
+	}
+	b.subcommands[name] = builder
```
Comment (pierluca) [inline cli/urfave/mod.go:88]: shouldn't we do error-checking to verify that `b.subcommands[name]` hasn't already been defined?

File: cli/urfave/mod_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	fakeAction := func(flags cli.Flags) error {
+		return nil
+	}
+
+	res = makeAction(fakeAction)
+	require.NotNil(t, res)
+
+	out := res(&ucli.Context{})
+	require.NoError(t, out)
```
Comment (pierluca) [inline cli/urfave/mod_test.go:117]: the logic can be severely broken without this test failing.
Can we do some "tougher" asserts about `res` and `out` ?

File: cli/urfave/mod_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		cli.IntFlag{
+			Name:     "name4",
+			Usage:    "usage4",
+			Required: true,
+			Value:    1,
+		},
+	}
+
+	out := buildFlags(in)
+	require.Len(t, out, 4)
```
Comment (pierluca) [inline cli/urfave/mod_test.go:93]: you might want to add a few assertions here with respect to what's really outputted.

File: cli/urfave/mod_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+func TestSetCommand(t *testing.T) {
+	builder := NewBuilder("test", nil)
+
+	builder.SetCommand("first")
+	builder.SetCommand("second")
+
+	app := builder.Build().(*ucli.App)
+
+	// urfave adds by default the -help command. This is why we expect 3.
+	require.Len(t, app.Commands, 3)
```
Comment (pierluca) [inline cli/urfave/mod_test.go:34]: can we partially replace the comment with some asserts about there being a "first", "second" and "help" ? :)

**Review by pierluca — CHANGES_REQUESTED** [review]
Few things here and there, nothing major :)

File: cli/urfave/mod.go (hunk 1 above)
Comment (nkcr) [reply cli/urfave/mod.go:3]: Even if the exposed functions are not urfave dependent, we name it after its underlying implementation. A user is then free to choose its implementation based on the implementation details, which can matter, like with could provide a `fake` cli. 
This is how it's done in Dela. This is why we, for example, have `minogrpc` and `minoch`, which both provides the same exported types but with different implementation details. Could say the same with the ordering module: we have a `pow (proof-of-work)` one and a `cosipbft` one.

File: cli/urfave/mod.go (hunk 2 above)
Comment (nkcr) [reply cli/urfave/mod.go:49]: That would be useful only if we needed to use that function from within the package and we needed to handle a `cmdBuilder`, which is not the case. An external user would see the abstraction anyway.

File: cli/urfave/mod.go (hunk 1 above)
Comment (pierluca) [reply cli/urfave/mod.go:3]: OK, I can accept this.
Could we at least have 'cli' in the package name, like for mino? It'd make the using code more readable (ex. `urcli.NewBuilder()` as opposed to `urfave.NewBuilder()`) 

For reference, I've tried to express my thinking below
---
Architecture is a way of supporting some desirable characteristics of a software system, in a general sense we're managing risks related to its evolution, stability and performance. As an example, we know that the overlay network implementation may need to change, so we design an abstraction for it (mino). We know that ordering will need to change, so we do the same. We want to be able to run experiments, so we make it easy to have multiple swappable implementations.

In the CLI case, do we really expect to have use for multiple implementations? A `fake` CLI seems something that belongs in an integration test package, for instance. On the other hand, does this mean that if we want to change the CLI implementation, we have to go through each and every user of the `urfave` package, to point to a different package in the dependency injection system ? That seems cumbersome and 'urfave' isn't quite explicit as a package name.

Ultimately, it's the CLI so I don't expect there'll be too many places making use of this, and this design might help change the urfave-based CLI programs one-at-a-time if need be. It makes sense.

File: cli/urfave/mod.go (hunk 2 above)
Comment (pierluca) [reply cli/urfave/mod.go:48]: agreed.
I was merely referring to the practice of returning concretions and letting the client code decide on the interface that suits its needs. In this case, I agree that it doesn't make much sense.

**Review by pierluca — APPROVED** [review]
I've added a couple of comments, but LGTM as-is.

**Pull request merged**